src/mappings/shared/flipperEventHandler.ts

```diff
@@ -0,0 +1,136 @@
+import { CallWith } from '../utils/types'
+import { BuyTokenEvent } from '../nfts/types'
+import {  pending, warn } from '../utils/logger'
+
+const OPERATION = 'BUY => FlipperEventHandler' as any
+
+type HandleFlipParams = {
+  event: CallWith<BuyTokenEvent>
+  nft: NFTEntity
+}
+
+export class FlipperEventHandler {
+  store: EntityManager
+
+  constructor(store: EntityManager) {
+    this.store = store
+  }
+
+  private getFlipperId(address: string, collectionId: string): string {
+    return `${address}-${collectionId}`
+  }
+
+  private getFlipEventId(address: string, collectionId: string, blockNumber: string): string {
+    return `${address}-${collectionId}-${blockNumber}`
+  }
+
+  async getOrCreateFlipper(address: string, collection: CollectionEntity, timestamp: Date): Promise<FlipperEntity> {
+    const id = this.getFlipperId(address, collection.id)
+    let flipper = await getOptional<FlipperEntity>(this.store, FlipperEntity, id)
+
+    if (!flipper) {
+      flipper = create(FlipperEntity, id, {
+        address,
+        collection,
+        owned: 0,
+        totalBought: BigInt(0),
+        totalSold: BigInt(0),
+        bestFlip: BigInt(0),
+        latestFlipTimestamp: timestamp,
+      })
+      await this.store.save(flipper)
+    }
+
+    return flipper
+  }
+
+  async handleNewOwner(event: CallWith<BuyTokenEvent>, nft: NFTEntity): Promise<FlipperEntity> {
+    const { caller: newOwner, collectionId, timestamp, price, blockNumber } = event
+    const flipper = await this.getOrCreateFlipper(newOwner, nft.collection, timestamp)
+
+    const flipEventId = this.getFlipEventId(newOwner, collectionId, blockNumber)
+    const flipEvent = create(FlipEvent, flipEventId, {
+      flipper,
+      nft,
+      boughtPrice: price,
+    })
+
+    await this.store.save(flipEvent)
+    await this.store.update(
+      FlipperEntity,
+      { id: flipper.id },
+      {
+        owned: flipper.owned + 1,
+        totalBought: BigInt(flipper.totalBought) + BigInt(price || 0),
+        latestFlipTimestamp: timestamp,
+      }
+    )
+
+    return flipper
+  }
+
+  async handlePreviousOwner(event: CallWith<BuyTokenEvent>, nft: NFTEntity): Promise<void> {
+    const { caller: newOwner, currentOwner: previousOwner, timestamp, price } = event
+
+    const previousFlipperId = this.getFlipperId(previousOwner, nft.collection.id)
+    let previousFlipper = await getOptional<FlipperEntity>(this.store, FlipperEntity, previousFlipperId)
+
+    if (!previousFlipper) {
+      pending(
+        OPERATION,
+        `No previous flipper found for the previous owner (${previousOwner}). No previous flip to handle.`
+      )
+      return
+    }
+
+    if (!previousFlipper.flips) {
+      warn(OPERATION, `No previous flips found for the previous owner (${previousOwner}). No previous flip to handle.`)
+      return
+    }
+
+    const previousFlipEvent = previousFlipper.flips.find((flip) => flip.nft.id === nft.id)
```

Comment:
What is the issue?

I am looking for the previous time this (current event) nft has been bought
if it has been bought before, then we are now dealing with a flip
otherwise, it is first time purchase, and the function `handlePreviousOwner` returns because the previous owner got the nft somehow (minting sending, whatever), but they didn't buy it => not a flip

Reply:
> What is the issue?

That iterating over array in javascript would be always slower than reading from DB. If the flips would have 10k items my bet is that processor would be slow and die 🤷

Reply:
I understand.
I'll query the db instead

👌👌

Reply:
@vikiival  please have a look again sir